gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/CleanableHiveDataset.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.thrift.TException;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.typesafe.config.Config;
+
+import gobblin.data.management.copy.hive.HiveDataset;
+import gobblin.data.management.policy.SelectBeforeTimeBasedPolicy;
+import gobblin.data.management.policy.VersionSelectionPolicy;
+import gobblin.data.management.version.HiveDatasetVersion;
+import gobblin.data.management.version.finder.DatePartitionHiveVersionFinder;
+import gobblin.data.management.version.finder.AbstractHiveDatasetVersionFinder;
+import gobblin.hive.HiveMetastoreClientPool;
+import gobblin.util.AutoReturnableObject;
+import gobblin.util.ConfigUtils;
+import gobblin.util.reflection.GobblinConstructorUtils;
+
+/**
+ * <p>
+ * A {@link HiveDataset} used for Retention. A {@link HiveDataset} represents a hive table and a {@link HiveDatasetVersion}
+ * represents a hive partition of this table.
+ * </p>
+ *
+ * <ul>
+ * <li>A version finder at {@value #VERSION_FINDER_CLASS_KEY} is used to find all the partitions the dataset
+ * <li>A selection policy at {@value #SELECTION_POLICY_CLASS_KEY} is appliced on all these partitions to get the partitions to be deleted.
+ * <li>These selected partitions are dropped in the hive metastore and all the data on FileSystem is also deleted
+ * </ul>
+ *
+ */
+@Slf4j
+@SuppressWarnings({ "rawtypes", "unchecked" })
+public class CleanableHiveDataset extends HiveDataset implements CleanableDataset {
+
+  public static final String SIMULATE_KEY = "gobblin.retention.simulate";
+  public static final String SIMULATE_DEFAULT = Boolean.toString(false);
+
+  public static final String SHOULD_DELETE_DATA_KEY = "gobblin.retention.hive.shouldDeleteData";
+  public static final String SHOULD_DELETE_DATA_DEFAULT = Boolean.toString(true);
+
+  public static final String VERSION_FINDER_CLASS_KEY = "version.finder.class";
+  public static final String DEFAULT_VERSION_FINDER_CLASS = DatePartitionHiveVersionFinder.class.getName();
+
+  public static final String SELECTION_POLICY_CLASS_KEY = "selection.policy.class";
+  public static final String DEFAULT_SELECTION_POLICY_CLASS = SelectBeforeTimeBasedPolicy.class.getName();
+
+  private final VersionSelectionPolicy hiveSelectionPolicy;
+  private final AbstractHiveDatasetVersionFinder hiveDatasetVersionFinder;
+
+  private final boolean simulate;
+  private final boolean shouldDeleteData;
+  private final FsCleanableHelper fsCleanableHelper;
+
+  public CleanableHiveDataset(FileSystem fs, HiveMetastoreClientPool clientPool, Table table, Properties jobProps, Config datasetConfig) throws IOException {
+    super(fs, clientPool, table, jobProps, datasetConfig);
+
+    try {
+      this.hiveSelectionPolicy =
+          (VersionSelectionPolicy) GobblinConstructorUtils.invokeFirstConstructor(
```

Comment:
Yes I agree. I think we should think through and come up with a nice builder that supports  ```invokeFirstConstructor```, ```invokeLongestConstructor``` etc.
Created an issue https://github.com/linkedin/gobblin/issues/1108